File: app/src/main/java/ru/e6atb/chat/MessageButtonRows.java
package ru.e6atb.chat;

import java.util.ArrayList;
import java.util.List;

final class MessageButtonRows {
	private MessageButtonRows() {
	}

	static List<List<MiniTaLib.Button>> group(List<MiniTaLib.Button> buttons) {
		ArrayList<List<MiniTaLib.Button>> rows = new ArrayList<List<MiniTaLib.Button>>();
		if (buttons == null) return rows;
		for (MiniTaLib.Button button : buttons) {
			if (button == null || button.text == null || button.text.length() == 0) continue;
			int rowIndex = Math.max(0, Math.min(11, button.row));
			while (rows.size() <= rowIndex) rows.add(null);
			List<MiniTaLib.Button> row = rows.get(rowIndex);
			if (row == null) {
				row = new ArrayList<MiniTaLib.Button>();
				rows.set(rowIndex, row);
			}
			row.add(button);
		}
		return rows;
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/MessageButtonRows.kt
@file:Suppress("EXPOSED_FUNCTION_RETURN_TYPE", "EXPOSED_PARAMETER_TYPE")

package ru.e6atb.chat

object MessageButtonRows {
    @JvmStatic fun group(buttons: List<MST5.Button>?): List<List<MST5.Button>?> {
        val rows = ArrayList<MutableList<MST5.Button>?>()
        buttons.orEmpty().forEach { button ->
            if (button.text.isNullOrEmpty()) return@forEach
            val index = button.row.coerceIn(0, 11)
            while (rows.size <= index) rows.add(null)
            val row = rows[index] ?: ArrayList<MST5.Button>().also { rows[index] = it }
            row.add(button)
        }
        return rows
    }
}
